### PR TITLE
Orca pytorch cifar

### PR DESCRIPTION
**Pull request by yaxinchen666**
Add different backends for orca pytorch cifar10 example.

### CODE REVIEW

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md [hunk 1]
```diff
@@ -14,14 +14,23 @@ pip install torchvision
 pip install jep==3.9.0
 pip install six cloudpickle
 pip install matplotlib
+pip install ray==1.2.0
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md:17]: Do not put the required packages altogether. What about directly referring to the documentation page https://analytics-zoo.readthedocs.io/en/latest/doc/Orca/QuickStart/orca-pytorch-distributed-quickstart.html#step-0-prepare-environment or separate the commands for bigdl/ray.

**Comment by hkvision** [comment]
@jason-dai Take a look if the changes are acceptable. Basically is the difference between instances and creators for bigdl/ray backends.

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md (hunk 1 above)
Comment (yaxinchen666) [reply pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md]: Modified. Please have a look : )

**Comment by hkvision** [comment]
Add a sentence in introduction, something like:

We demonstrate two distributed PyTorch training backends for this example, namely "bigdl" and "torch_distributed". You can run with either backend as you wish.

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md [hunk 2]
```diff
@@ -11,17 +11,28 @@ conda activate zoo
 pip install analytics-zoo  # 0.10.0.dev3 or above
 pip install torch
 pip install torchvision
+pip install matplotlib
+# for bigdl:
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md:15]: add an empty line before this line

Comment (hkvision) [reply pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md:15]: Change to `For bigdl backend`

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md [hunk 3]
```diff
@@ -11,17 +11,28 @@ conda activate zoo
...
 pip install jep==3.9.0
 pip install six cloudpickle
-pip install matplotlib
+# for torch_distributed:
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md:18]: For torch_distributed backend

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md [hunk 4]
```diff
@@ -11,17 +11,28 @@ conda activate zoo
...
+pip install ray==1.2.0
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md:19]: pip install analytics-zoo[ray]

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md [hunk 5]
```diff
@@ -31,6 +42,8 @@ python cifar10.py --cluster_mode yarn-client
 
 ## Results
 
+### With bigdl backend
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md:45]: remove this subtitle

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md [hunk 6]
```diff
@@ -31,6 +42,8 @@ python cifar10.py --cluster_mode yarn-client
...
+
 You can find the logs for training as follows:
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md:47]: For bigdl backend, you can...

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py [hunk 7]
```diff
@@ -39,6 +39,8 @@
 parser = argparse.ArgumentParser(description='PyTorch Cifar10 Example')
 parser.add_argument('--cluster_mode', type=str, default="local",
                     help='The cluster mode, such as local, yarn or k8s.')
+parser.add_argument('--backend', type=str, default="bigdl",
+                    help='The backend of torch Estimator, such as bigdl, torch-distributed or horovod')
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:43]: torch -> PyTorch
bigdl and torch_distributed are supported.

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py [hunk 8]
```diff
@@ -53,16 +55,24 @@
 transform = transforms.Compose(
     [transforms.ToTensor(),
      transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])
+root_dir = './data'
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:58]: put root_dir into config?

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py [hunk 9]
```diff
@@ -104,24 +105,63 @@ def forward(self, x):
...
+    optimizer = optim.SGD(model.parameters(), 
+                          lr=config.get("lr", 0.001),
+                          momentum=config.get("momentum", 0.9))
+    return optimizer
+
+
 criterion = nn.CrossEntropyLoss()
-optimizer = optim.SGD(net.parameters(), lr=0.001, momentum=0.9)
-
-net.train()
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:111]: do we need to put net.train() in model_creator?

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py [hunk 10]
```diff
@@ -104,24 +105,63 @@ def forward(self, x):
...
+
+    res = orca_estimator.evaluate(data=test_loader)
+    print("Accuracy of the network on the test images: %s" % res)
+elif args.backend == "torch_distributed":
+    orca_estimator = Estimator.from_torch(model=model_creator,
+                                          optimizer=optim_creator,
+                                          loss=criterion,
+                                          metrics=[Accuracy()],
+                                          backend="torch_distributed",
+                                          config={"lr": 0.001})
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:159]: add an else branch and raise an error like: "Only bigdl and torch_distributed are supported as the backend, but got {}".format(args.backend)

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py [hunk 11]
```diff
@@ -104,24 +105,63 @@ def forward(self, x):
...
-                                      metrics=[Accuracy()],
-                                      backend="bigdl")
-orca_estimator.fit(data=trainloader, epochs=2, validation_data=testloader,
-                   checkpoint_trigger=EveryEpoch())
-print('Finished Training')
-dataiter = iter(testloader)
+batch_size = 4
+
+# plot some random images
+dataiter = iter(train_loader_creator(config={}, batch_size=batch_size))
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:124]: just make train_loader a variable here so that you don't need to call again for bigdl backend

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py (hunk 9 above)
Comment (yaxinchen666) [reply pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:111]: https://pytorch.org/docs/stable/generated/torch.nn.Module.html
net.train() is to switch the training mode and only has effect when there is Dropout/BatchNorm in the model. I think it makes no difference here so I just remove it.

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md [hunk 12]
```diff
@@ -8,20 +8,32 @@ We recommend you to use Anaconda to prepare the environment, especially if you w
...
 pip install torch
 pip install torchvision
+pip install matplotlib
+
+# For bigdl backend:
+pip install analytics-zoo  # 0.10.0.dev3 or above
 pip install jep==3.9.0
 pip install six cloudpickle
-pip install matplotlib
+# For torch_distributed backend:
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md:19]: add a blank line between them

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md [hunk 13]
```diff
@@ -1,5 +1,5 @@
 # PyTorch Cifar10 example
-We demonstrate how to easily run synchronous distributed PyTorch training using PyTorch Estimator of Project Orca in Analytics Zoo. We use a simple convolutional nueral network model to train on Cifar10 dataset, which is a dataset for image classification. See [here](https://pytorch.org/tutorials/beginner/blitz/cifar10_tutorial.html) for the original single-node version of this example provided by PyTorch.
+We demonstrate how to easily run synchronous distributed PyTorch training using PyTorch Estimator of Project Orca in Analytics Zoo. We use a simple convolutional nueral network model to train on Cifar10 dataset, which is a dataset for image classification. See [here](https://pytorch.org/tutorials/beginner/blitz/cifar10_tutorial.html) for the original single-node version of this example provided by PyTorch. We demonstrate two distributed PyTorch training backends for this example, namely "bigdl" and "torch_distributed". You can run with either backend as you wish.
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md:2]: demonstrate -> provide?

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md [hunk 14]
```diff
@@ -31,6 +43,8 @@ python cifar10.py --cluster_mode yarn-client
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/README.md:44]: Add a sentence like `The default backend is `bigdl`. You can also run with `torch_distributed` by specifying the backend.`

File: pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py [hunk 15]
```diff
@@ -39,6 +39,10 @@
 parser = argparse.ArgumentParser(description='PyTorch Cifar10 Example')
 parser.add_argument('--cluster_mode', type=str, default="local",
                     help='The cluster mode, such as local, yarn or k8s.')
+parser.add_argument('--backend', type=str, default="bigdl",
+                    help='The backend of PyTorch Estimator; bigdl and torch-distributed are supported')
+parser.add_argument('--root_dir', type=str, default="./data",
```
Comment (hkvision) [inline pyzoo/zoo/examples/orca/learn/pytorch/cifar10/cifar10.py:44]: I'm not meaning make this as an argument, but pass config={"root_dir": "./data"} when creating the Estimator...

**Comment by hkvision** [comment]
Need to add the test for torch_distributed in run-example-test-ray.sh and run-example-test-pip-ray.sh

**Review by hkvision — APPROVED** [review]
LGTM. Can merge after jenkins passed. Need to pay attention to the jenkins results of the added examples.

**Comment by yaxinchen666** [comment]
Jenkins:
http://10.239.47.210:18888/job/ZOO-PR-Validation/5319/
ZOO-PR-Python-Spark-2.4-py36-ray & ZOO-PR-Python-Spark-2.4-py37-ray failed. 
Rerun and passed:
http://10.239.47.210:18888/job/ZOO-PR-Python-Spark-2.4-py36-ray/1539/
http://10.239.47.210:18888/job/ZOO-PR-Python-Spark-2.4-py37-ray/1562/

**Comment by hkvision** [comment]
Paste the link for running example job of ray, and check the logs to make sure this example has passed.

**Comment by yaxinchen666** [comment]
> Paste the link for running example job of ray, and check the logs to make sure this example has passed.

http://10.239.47.210:18888/job/ZOO-PR-Python-ExampleTests/2546/

`#8 orca cifar10 example time used:210 seconds` printed in
http://10.239.47.210:18888/job/ZOO-PR-Python-ExampleTests-py37-ray/1560/consoleFull
